src/vm/builtins/builtin_runner/range_check.zig

```diff
@@ -43,26 +63,237 @@ pub const RangeCheckBuiltinRunner = struct {
+    /// with the Range Check runner's memory segments. The `stop_ptr` may be `null`.
+    ///
+    /// # Returns
+    /// A tuple of `usize` and `?usize` addresses.
+    pub fn getMemorySegmentAddresses(self: *Self) std.meta.Tuple(&.{
+        usize,
+        ?usize,
+    }) {
+        return .{
+            self.base,
+            self.stop_ptr,
+        };
+    }
+
+    /// Calculate the final stack.
+    ///
+    /// This function calculates the final stack pointer for the Range Check runner, based on the provided `segments`, `pointer`, and `self` settings. If the runner is included,
+    /// it verifies the stop pointer for consistency and sets it. Otherwise, it sets the stop pointer to zero.
+    ///
+    /// # Parameters
+    ///
+    /// - `segments`: A pointer to the `MemorySegmentManager` for segment management.
+    /// - `pointer`: A `Relocatable` pointer to the current stack pointer.
+    ///
+    /// # Returns
+    ///
+    /// A `Relocatable` pointer to the final stack pointer, or an error code if the
+    /// verification fails.
+    pub fn finalStack(
+        self: *Self,
+        segments: *MemorySegmentManager,
+        pointer: Relocatable,
+    ) !Relocatable {
+        if (self.included) {
+            const stop_pointer_addr = pointer.subUint(
+                @intCast(1),
+            ) catch return RunnerError.NoStopPointer;
+            const stop_pointer = try (segments.memory.get(
+                stop_pointer_addr,
+            ) catch return RunnerError.NoStopPointer).tryIntoRelocatable();
+            if (@as(
+                isize,
+                @intCast(self.base),
+            ) != stop_pointer.segment_index) {
+                return RunnerError.InvalidStopPointerIndex;
+            }
+            const stop_ptr = stop_pointer.offset;
+
+            if (stop_ptr != try self.getUsedInstances(segments) * @as(
+                usize,
+                @intCast(self.cells_per_instance),
+            )) {
+                return RunnerError.InvalidStopPointer;
+            }
+            self.stop_ptr = stop_ptr;
+            return stop_pointer_addr;
+        }
+
+        self.stop_ptr = 0;
+        return pointer;
+    }
+
+    /// Creates Validation Rules ArrayList
+    ///
+    /// # Parameters
+    ///
+    /// - `memory`: A `Memory` pointer of validation rules segment index.
+    /// - `address`: A `Relocatable` pointer to the validation rule.
+    ///
+    /// # Returns
+    ///
+    /// An `ArrayList(Relocatable)` containing the rules address
+    /// verification fails.
+    pub fn rangeCheckValidationRule(memory: *Memory, address: Relocatable) std.ArrayList(!Relocatable) {
+        const addr = memory.get(address);
+        if (addr.bits <= N_PARTS * INNER_RC_BOUND_SHIFT) {
+            return std.ArrayList.append(address);
+        } else {
+            return std.ArrayList.append(Error.MemoryOutOfBounds);
+        }
+    }
```

Comment:
I'm not sure this will work as expect because your `ArrayList` is not initialized before `append` method is used. See for example: https://github.com/ziglang/zig/blob/e74ced21b70f229167765d863613d3d5033b93fe/lib/std/array_list.zig#L1188-L1191